File: notes/section2/run/s2p4_hello.py
# STAT 5400 - Section 2.4 - your first LLM API call (Python, openai)
# The key is a Codespaces secret. Never write it in this file.
# Check the OpenAI key, and on IDAS use the class copy of the openai package
import os, sys, urllib.request
if os.path.isdir(os.path.expanduser("~/classdata/models/python")):          # on IDAS
    sys.path.insert(0, os.path.expanduser("~/classdata/models/python"))     # class packages first
    sys.modules.pop("typing_extensions", None)                              # forget the old system copy
renviron = os.path.expanduser("~/.Renviron")                                # read the key from the file R uses
if os.path.exists(renviron):
    for line in open(renviron):
        name, _, value = line.strip().partition("=")
        if name and not name.startswith("#") and value:
            os.environ.setdefault(name, value.strip("\"'"))
help_url = "https://boxiang-wang.github.io/stat5400/notes/section2/s2p4.html#setting-up-idas"
if not os.environ.get("OPENAI_API_KEY"):
    print("No OpenAI key found. How to set it up:", help_url)
else:
    try:
        urllib.request.urlopen(urllib.request.Request("https://api.openai.com/v1/models",
            headers={"Authorization": "Bearer " + os.environ["OPENAI_API_KEY"]}), timeout=10)
    except OSError:
        print("Your OpenAI key did not work. Check it, or see:", help_url)

import os
from openai import OpenAI

client = OpenAI()

response = client.responses.create(
    model="gpt-5-nano",
    input=[{"role": "user", "content": "Write a one-sentence bedtime story about a unicorn."}],
)
print(response.output_text)

# Every request is stateless: a second, separate call remembers nothing.
r1 = client.responses.create(
    model="gpt-5-nano",
    input=[{"role": "user", "content": "Write a paragraph about the Department of Statistics "
                                       "and Actuarial Science at the University of Iowa."}])
print(r1.output_text)

r2 = client.responses.create(
    model="gpt-5-nano",
    input=[{"role": "user", "content": "What's the name of the university that I just mentioned?"}])
print(r2.output_text)        # it does not know

# Three ways to give it a memory. This is the third: let the provider keep the thread.
r3 = client.responses.create(
    model="gpt-5-nano",
    input=[{"role": "user", "content": "Write two sentences about the University of Iowa."}],
    store=True)
print(r3.output_text)

r4 = client.responses.create(
    model="gpt-5-nano",
    previous_response_id=r3.id,
    input=[{"role": "user", "content": "What is the name of the university I just mentioned?"}],
    store=True)
print(r4.output_text)

File: notes/section2/run/s2p4_hello.R
# STAT 5400 - Section 2.4 - your first LLM API call (R, ellmer)
# The key is a Codespaces secret. Never write it in this file.
# Check the OpenAI key (reads ~/.Renviron again, so no restart is needed after you add the key)
if (file.exists("~/.Renviron")) readRenviron("~/.Renviron")
help_url <- "https://boxiang-wang.github.io/stat5400/notes/section2/s2p4.html#setting-up-idas"
if (!nzchar(Sys.getenv("OPENAI_API_KEY"))) {
  message("No OpenAI key found. How to set it up: ", help_url)
} else if (inherits(try(suppressWarnings(readLines(url("https://api.openai.com/v1/models",
             headers = c(Authorization = paste("Bearer", Sys.getenv("OPENAI_API_KEY")))))), silent = TRUE), "try-error")) {
  message("Your OpenAI key did not work. Check it, or see: ", help_url)
}

# Start the Ollama server if it is not running yet (needed once per IDAS session)
ollama_up <- function() !inherits(try(suppressWarnings(readLines("http://localhost:11434/api/version")), silent = TRUE), "try-error")
if (!ollama_up()) {
  if (dir.exists("~/classdata/models")) {             # on IDAS: use the class copy of Ollama and its models
    Sys.setenv(PATH = paste0(path.expand("~/classdata/models/ollama/bin:"), Sys.getenv("PATH")),
               OLLAMA_MODELS = path.expand("~/classdata/models/library"), OLLAMA_NOPRUNE = "1")
  }
  if (nzchar(Sys.which("ollama"))) {                  # skip if Ollama is not installed
    system("nohup ollama serve > ~/ollama.log 2>&1 &")  # start it in the background
    for (i in 1:30) if (ollama_up()) break else Sys.sleep(1)
  }
}

library(ellmer)

chat <- chat_openai(model = "gpt-5-nano")     # reads OPENAI_API_KEY
chat$chat("Write a one-sentence bedtime story about a unicorn.")

# Run it again. The answer is different, because the model samples.
chat2 <- chat_openai(model = "gpt-5-nano")
chat2$chat("Write a one-sentence bedtime story about a unicorn.")

# A chat object keeps the conversation, so the second turn remembers the first.
chat3 <- chat_openai(model = "gpt-5-nano")
chat3$chat("Write two sentences about the University of Iowa.")
chat3$chat("What is the name of the university I just mentioned?")

# What it cost, and which model answered.
chat3$get_model()
chat3$get_tokens()

# Another provider, same code: only the address and the model name change.
# The small local model needs no key and runs inside the Codespace.
local <- chat_ollama(model = "llama3.2:1b")
local$chat("Give me 3 distributions in the exponential family.")
